
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/pheval/api/pheval/utils/phenopacket_utils/">
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-8.5.10">
    
    
      
        <title>Phenopacket utils - PhEval</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.975780f9.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.2505c338.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/mermaid@8.5.1/dist/mermaid.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.pheval.utils.phenopacket_utils" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="PhEval" class="md-header__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PhEval
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Phenopacket utils
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Night Mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Night Mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
            </label>
          
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Day Mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Day Mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="PhEval" class="md-nav__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    PhEval
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../pipeline/" class="md-nav__link">
        PhEval Pipeline
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../contact/" class="md-nav__link">
        Contact Us
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          API Documentation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API Documentation" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          API Documentation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/" class="md-nav__link">
        Cli
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../config_parser/" class="md-nav__link">
        Config parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../constants/" class="md-nav__link">
        Constants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../run_metadata/" class="md-nav__link">
        Run metadata
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_5" type="checkbox" id="__nav_5_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_5">
          analyse
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="analyse" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_5">
          <span class="md-nav__icon md-icon"></span>
          analyse
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/analysis/" class="md-nav__link">
        Analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/generate_plots/" class="md-nav__link">
        Generate plots
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/generate_summary_outputs/" class="md-nav__link">
        Generate summary outputs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/rank_stats/" class="md-nav__link">
        Rank stats
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_6" type="checkbox" id="__nav_5_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_6">
          post_processing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="post_processing" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_6">
          <span class="md-nav__icon md-icon"></span>
          post_processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../post_processing/post_processing/" class="md-nav__link">
        Post processing
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_7" type="checkbox" id="__nav_5_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_7">
          prepare
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="prepare" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_7">
          <span class="md-nav__icon md-icon"></span>
          prepare
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/create_noisy_phenopackets/" class="md-nav__link">
        Create noisy phenopackets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/create_spiked_vcf/" class="md-nav__link">
        Create spiked vcf
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/custom_exceptions/" class="md-nav__link">
        Custom exceptions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/update_phenopacket/" class="md-nav__link">
        Update phenopacket
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_8" type="checkbox" id="__nav_5_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_8">
          runners
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="runners" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_8">
          <span class="md-nav__icon md-icon"></span>
          runners
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../runners/runner/" class="md-nav__link">
        Runner
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_9" type="checkbox" id="__nav_5_9" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5_9">
          utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="utils" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_9">
          <span class="md-nav__icon md-icon"></span>
          utils
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../file_utils/" class="md-nav__link">
        File utils
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Phenopacket utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Phenopacket utils
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils" class="md-nav__link">
    src.pheval.utils.phenopacket_utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="md-nav__link">
    GeneIdentifierUpdater
  </a>
  
    <nav class="md-nav" aria-label="GeneIdentifierUpdater">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="md-nav__link">
    find_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="md-nav__link">
    obtain_gene_symbol_from_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="md-nav__link">
    update_genomic_interpretations_gene_identifier()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="md-nav__link">
    IncompatibleGenomeAssemblyError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="md-nav__link">
    PhenopacketRebuilder
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketRebuilder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="md-nav__link">
    add_randomised_hpo()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="md-nav__link">
    add_spiked_vcf_path()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="md-nav__link">
    update_interpretations()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="md-nav__link">
    causative_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="md-nav__link">
    diagnosed_genes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="md-nav__link">
    diagnosed_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="md-nav__link">
    files()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="md-nav__link">
    interpretations()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="md-nav__link">
    negated_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="md-nav__link">
    observed_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="md-nav__link">
    phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="md-nav__link">
    sample_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="md-nav__link">
    vcf_file_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="md-nav__link">
    create_hgnc_dict()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_json_message" class="md-nav__link">
    create_json_message()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.phenopacket_reader" class="md-nav__link">
    phenopacket_reader()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.write_phenopacket" class="md-nav__link">
    write_phenopacket()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../semsim_utils/" class="md-nav__link">
        Semsim utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        Utils
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Resources for contributors
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Resources for contributors" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Resources for contributors
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../developing_a_pheval_plugin/" class="md-nav__link">
        Developing a PhEval Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../contributing/" class="md-nav__link">
        Contributions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../styleguide/" class="md-nav__link">
        Monarch Style Guide for PhEval
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../CODE_OF_CONDUCT/" class="md-nav__link">
        Contributor Covenant Code of Conduct
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../roadmap/" class="md-nav__link">
        Roadmap
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils" class="md-nav__link">
    src.pheval.utils.phenopacket_utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="md-nav__link">
    GeneIdentifierUpdater
  </a>
  
    <nav class="md-nav" aria-label="GeneIdentifierUpdater">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="md-nav__link">
    find_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="md-nav__link">
    obtain_gene_symbol_from_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="md-nav__link">
    update_genomic_interpretations_gene_identifier()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="md-nav__link">
    IncompatibleGenomeAssemblyError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="md-nav__link">
    PhenopacketRebuilder
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketRebuilder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="md-nav__link">
    add_randomised_hpo()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="md-nav__link">
    add_spiked_vcf_path()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="md-nav__link">
    update_interpretations()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="md-nav__link">
    causative_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="md-nav__link">
    diagnosed_genes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="md-nav__link">
    diagnosed_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="md-nav__link">
    files()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="md-nav__link">
    interpretations()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="md-nav__link">
    negated_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="md-nav__link">
    observed_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="md-nav__link">
    phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="md-nav__link">
    sample_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="md-nav__link">
    vcf_file_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="md-nav__link">
    create_hgnc_dict()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_json_message" class="md-nav__link">
    create_json_message()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.phenopacket_reader" class="md-nav__link">
    phenopacket_reader()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.write_phenopacket" class="md-nav__link">
    write_phenopacket()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/monarch-initiative/pheval/edit/docs/docs/api/pheval/utils/phenopacket_utils.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>Phenopacket utils</h1>

<div class="doc doc-object doc-module">


<a id="src.pheval.utils.phenopacket_utils"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="doc doc-heading">
        <code>GeneIdentifierUpdater</code>


</h2>


  <div class="doc doc-contents ">



        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GeneIdentifierUpdater</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hgnc_data</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">identifier_map</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span> <span class="o">=</span> <span class="n">hgnc_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span> <span class="o">=</span> <span class="n">gene_identifier</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span> <span class="o">=</span> <span class="n">identifier_map</span>

    <span class="k">def</span> <span class="nf">find_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Finds the specified gene identifier for a gene symbol.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">_symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                        <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">query_gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtain gene symbol from a gene identifier. (e.g.)</span>
<span class="sd">        &quot;</span>
<span class="sd">        obtain_gene_symbol_from_identifier(query_gene_identifier=&quot;HGNC:5&quot;)</span>
<span class="sd">        &quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span><span class="p">[</span><span class="n">query_gene_identifier</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">_find_alternate_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Finds the alternate IDs for a gene symbol.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;ncbigene:&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;ensembl:&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;symbol:&quot;</span> <span class="o">+</span> <span class="n">gene_symbol</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                        <span class="k">return</span> <span class="p">[</span>
                            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;ncbigene:&quot;</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;ensembl:&quot;</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;symbol:&quot;</span> <span class="o">+</span> <span class="n">symbol</span><span class="p">,</span>
                        <span class="p">]</span>

    <span class="k">def</span> <span class="nf">update_genomic_interpretations_gene_identifier</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Updates the genomic interpretations of a phenopacket.&quot;&quot;&quot;</span>
        <span class="n">updated_interpretations</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">interpretations</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">updated_interpretation</span> <span class="ow">in</span> <span class="n">updated_interpretations</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">updated_interpretation</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">find_identifier</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">del</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="p">[:]</span>
                <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_find_alternate_ids</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                    <span class="p">)</span>
                <span class="p">)</span>
        <span class="k">return</span> <span class="n">updated_interpretations</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">find_identifier</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Finds the specified gene identifier for a gene symbol.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Finds the specified gene identifier for a gene symbol.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="n">query_gene_identifier</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Obtain gene symbol from a gene identifier. (e.g.)
"
obtain_gene_symbol_from_identifier(query_gene_identifier="HGNC:5")
"</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">query_gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Obtain gene symbol from a gene identifier. (e.g.)</span>
<span class="sd">    &quot;</span>
<span class="sd">    obtain_gene_symbol_from_identifier(query_gene_identifier=&quot;HGNC:5&quot;)</span>
<span class="sd">    &quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span><span class="p">[</span><span class="n">query_gene_identifier</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">update_genomic_interpretations_gene_identifier</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Updates the genomic interpretations of a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_genomic_interpretations_gene_identifier</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Updates the genomic interpretations of a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">updated_interpretations</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">interpretations</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">updated_interpretation</span> <span class="ow">in</span> <span class="n">updated_interpretations</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">updated_interpretation</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">find_identifier</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">del</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="p">[:]</span>
            <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_find_alternate_ids</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">updated_interpretations</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="doc doc-heading">
        <code>IncompatibleGenomeAssemblyError</code>


</h2>


  <div class="doc doc-contents ">
      <p class="doc doc-class-bases">
        Bases: <code>Exception</code></p>

  
      <p>Exception raised for incompatible genome assembly.</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">IncompatibleGenomeAssemblyError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Exception raised for incompatible genome assembly.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Incompatible Genome Assembly&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">assembly</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">phenopacket</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">message</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="si">}</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="doc doc-heading">
        <code>PhenopacketRebuilder</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Rebuilds a Phenopacket.</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">PhenopacketRebuilder</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Rebuilds a Phenopacket.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">:</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span> <span class="o">=</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">update_interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds the updated interpretations to a phenopacket.&quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">add_randomised_hpo</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">randomised_hpo</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds randomised phenotypic profile to phenopacket.&quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">add_spiked_vcf_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spiked_vcf_file_data</span><span class="p">:</span> <span class="n">File</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds spiked vcf path to phenopacket.&quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="n">phenopacket_files</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;vcf&quot;</span>
        <span class="p">]</span>
        <span class="n">phenopacket_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">phenopacket_files</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="doc doc-heading">
<code class="highlight language-python"><span class="n">add_randomised_hpo</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Adds randomised phenotypic profile to phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_randomised_hpo</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">randomised_hpo</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds randomised phenotypic profile to phenopacket.&quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="doc doc-heading">
<code class="highlight language-python"><span class="n">add_spiked_vcf_path</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Adds spiked vcf path to phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_spiked_vcf_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spiked_vcf_file_data</span><span class="p">:</span> <span class="n">File</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds spiked vcf path to phenopacket.&quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="n">phenopacket_files</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;vcf&quot;</span>
    <span class="p">]</span>
    <span class="n">phenopacket_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="p">[:]</span>
    <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">phenopacket_files</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="doc doc-heading">
<code class="highlight language-python"><span class="n">update_interpretations</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Adds the updated interpretations to a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the updated interpretations to a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="doc doc-heading">
        <code>PhenopacketUtil</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Retrieves relevant data from a phenopacket.</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">PhenopacketUtil</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves relevant data from a phenopacket.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_contents</span><span class="p">:</span> <span class="n">Phenopacket</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span> <span class="o">=</span> <span class="n">phenopacket_contents</span>

    <span class="k">def</span> <span class="nf">sample_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Retrieve the sample ID from a phenopacket or proband of a family.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span> <span class="nf">phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Retrieves a list of all HPO terms.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">observed_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Removes any HPO terms labelled as excluded.&quot;&quot;&quot;</span>
        <span class="n">phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">negated_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Retrieve negated phenotypic features.&quot;&quot;&quot;</span>
        <span class="n">negated_phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="n">negated_phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">negated_phenotypic_features</span>

    <span class="k">def</span> <span class="nf">interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns all interpretations of a phenopacket.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">interpretations</span>

    <span class="k">def</span> <span class="nf">causative_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">ProbandCausativeVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of all causative variants listed in a phenopacket.&quot;&quot;&quot;</span>
        <span class="n">all_variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span>
                <span class="n">genotype</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">allelic_state</span>
                <span class="n">variant_data</span> <span class="o">=</span> <span class="n">ProbandCausativeVariant</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">genome_assembly</span><span class="p">,</span>
                    <span class="n">GenomicVariant</span><span class="p">(</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">genotype</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">all_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">all_variants</span>

    <span class="k">def</span> <span class="nf">files</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns all files associated with a phenopacket.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">files</span>

    <span class="k">def</span> <span class="nf">vcf_file_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">File</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Retrieves the genome assembly and vcf name from a phenopacket.&quot;&quot;&quot;</span>
        <span class="n">compatible_genome_assembly</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">]</span>
        <span class="n">vcf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">()</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;vcf&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span>
            <span class="s2">&quot;.vcf.gz&quot;</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="n">IncorrectFileFormatError</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">),</span> <span class="s2">&quot;.vcf or .vcf.gz file&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">compatible_genome_assembly</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">IncompatibleGenomeAssemblyError</span><span class="p">(</span>
                <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">],</span> <span class="n">phenopacket_path</span>
            <span class="p">)</span>
        <span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">vcf_data</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_extract_diagnosed_gene</span><span class="p">(</span>
        <span class="n">genomic_interpretation</span><span class="p">:</span> <span class="n">GenomicInterpretation</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProbandCausativeGene</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the disease causative gene from the variant descriptor field if not empty,</span>
<span class="sd">        otherwise, returns from the gene descriptor from a phenopacket.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">ByteSize</span><span class="p">()</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">gene_symbol</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">gene_identifier</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">diagnosed_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">ProbandCausativeGene</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a unique list of all causative genes and the corresponding gene identifiers from a phenopacket.&quot;&quot;&quot;</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_diagnosed_gene</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
                <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">}</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">genes</span>

    <span class="k">def</span> <span class="nf">diagnosed_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">GenomicVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of all variants from a phenopacket - for use in assess-prioritisation.&quot;&quot;&quot;</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">variant</span> <span class="o">=</span> <span class="n">GenomicVariant</span><span class="p">(</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">pos</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">ref</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                    <span class="n">alt</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="doc doc-heading">
<code class="highlight language-python"><span class="n">causative_variants</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a list of all causative variants listed in a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">causative_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">ProbandCausativeVariant</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of all causative variants listed in a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">all_variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span>
            <span class="n">genotype</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">allelic_state</span>
            <span class="n">variant_data</span> <span class="o">=</span> <span class="n">ProbandCausativeVariant</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="n">vcf_record</span><span class="o">.</span><span class="n">genome_assembly</span><span class="p">,</span>
                <span class="n">GenomicVariant</span><span class="p">(</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">genotype</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">all_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant_data</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">all_variants</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagnosed_genes</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a unique list of all causative genes and the corresponding gene identifiers from a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagnosed_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">ProbandCausativeGene</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a unique list of all causative genes and the corresponding gene identifiers from a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_diagnosed_gene</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">}</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">genes</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagnosed_variants</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a list of all variants from a phenopacket - for use in assess-prioritisation.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagnosed_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">GenomicVariant</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of all variants from a phenopacket - for use in assess-prioritisation.&quot;&quot;&quot;</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">variant</span> <span class="o">=</span> <span class="n">GenomicVariant</span><span class="p">(</span>
                <span class="n">chrom</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                <span class="n">pos</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                <span class="n">ref</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                <span class="n">alt</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="doc doc-heading">
<code class="highlight language-python"><span class="n">files</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns all files associated with a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">files</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns all files associated with a phenopacket.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">files</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="doc doc-heading">
<code class="highlight language-python"><span class="n">interpretations</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns all interpretations of a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns all interpretations of a phenopacket.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">interpretations</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">negated_phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve negated phenotypic features.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">negated_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve negated phenotypic features.&quot;&quot;&quot;</span>
    <span class="n">negated_phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
            <span class="n">negated_phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">negated_phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">observed_phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Removes any HPO terms labelled as excluded.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">observed_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Removes any HPO terms labelled as excluded.&quot;&quot;&quot;</span>
    <span class="n">phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieves a list of all HPO terms.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves a list of all HPO terms.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="doc doc-heading">
<code class="highlight language-python"><span class="n">sample_id</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve the sample ID from a phenopacket or proband of a family.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sample_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve the sample ID from a phenopacket or proband of a family.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="doc doc-heading">
<code class="highlight language-python"><span class="n">vcf_file_data</span><span class="p">(</span><span class="n">phenopacket_path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieves the genome assembly and vcf name from a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">vcf_file_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">File</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves the genome assembly and vcf name from a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">compatible_genome_assembly</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">]</span>
    <span class="n">vcf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">()</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;vcf&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span>
        <span class="s2">&quot;.vcf.gz&quot;</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="n">IncorrectFileFormatError</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">),</span> <span class="s2">&quot;.vcf or .vcf.gz file&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">compatible_genome_assembly</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">IncompatibleGenomeAssemblyError</span><span class="p">(</span>
            <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">],</span> <span class="n">phenopacket_path</span>
        <span class="p">)</span>
    <span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">vcf_data</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="doc doc-heading">
<code class="highlight language-python"><span class="n">create_hgnc_dict</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Creates reference for updating gene symbols and identifiers.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_hgnc_dict</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">defaultdict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates reference for updating gene symbols and identifiers.&quot;&quot;&quot;</span>
    <span class="n">hgnc_df</span> <span class="o">=</span> <span class="n">read_hgnc_data</span><span class="p">()</span>
    <span class="n">hgnc_data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">_index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">hgnc_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">previous_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;refseq_accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;refseq_accession&quot;</span><span class="p">]</span>
        <span class="n">previous</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;prev_symbol&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">previous</span><span class="p">:</span>
            <span class="n">previous_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">))</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">previous_names</span>

    <span class="k">return</span> <span class="n">hgnc_data</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.create_json_message" class="doc doc-heading">
<code class="highlight language-python"><span class="n">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Creates json message for writing to file.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates json message for writing to file.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">MessageToJson</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.phenopacket_reader" class="doc doc-heading">
<code class="highlight language-python"><span class="n">phenopacket_reader</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Reads a phenopacket file, returning its contents.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">phenopacket_reader</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n">Path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reads a phenopacket file, returning its contents.&quot;&quot;&quot;</span>
    <span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;proband&quot;</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Parse</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">),</span> <span class="n">Family</span><span class="p">())</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Parse</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">),</span> <span class="n">Phenopacket</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.write_phenopacket" class="doc doc-heading">
<code class="highlight language-python"><span class="n">write_phenopacket</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Writes a phenopacket.</p>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_phenopacket</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Phenopacket</span> <span class="ow">or</span> <span class="n">Family</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Writes a phenopacket.&quot;&quot;&quot;</span>
    <span class="n">phenopacket_json</span> <span class="o">=</span> <span class="n">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">phenopacket_json</span><span class="p">)</span>
    <span class="n">outfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../file_utils/" class="md-footer__link md-footer__link--prev" aria-label="Previous: File utils" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              File utils
            </div>
          </div>
        </a>
      
      
        
        <a href="../semsim_utils/" class="md-footer__link md-footer__link--next" aria-label="Next: Semsim utils" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Semsim utils
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../../..", "features": [], "search": "../../../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.5a2dcb6a.min.js"></script>
      
        <script src="https://unpkg.com/mermaid@8.5.1/dist/mermaid.min.js"></script>
      
    
    
  </body>
</html>